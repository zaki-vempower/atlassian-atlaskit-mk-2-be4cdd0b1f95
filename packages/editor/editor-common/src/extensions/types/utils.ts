import { Icon, MaybeADFEntity } from './extension-manifest';

export type MenuItem = {
  key: string;
  title: string;
  description: string;
  icon: Icon;
  node: MaybeADFEntity;
};

export type MenuItemMap = { [key: string]: MenuItem };
