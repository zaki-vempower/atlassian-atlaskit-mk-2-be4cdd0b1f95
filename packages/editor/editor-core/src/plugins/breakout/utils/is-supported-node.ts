import { Node as PMNode } from 'prosemirror-model';

const supportedNodesForBreakout = ['codeBlock', 'layoutSection', 'expand'];

export function isSupportedNodeForBreakout(node: PMNode): boolean {
  return supportedNodesForBreakout.indexOf(node.type.name) !== -1;
}
