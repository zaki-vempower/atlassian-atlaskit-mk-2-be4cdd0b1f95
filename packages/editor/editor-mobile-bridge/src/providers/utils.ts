import { createPromise } from '../cross-platform-promise';

const globalFetch = window.fetch;
export const mockFetchFor = (urls: Array<string> = []) => {
  window.fetch = (request, options) => {
    let url = typeof request === 'string' ? request : request.url;

    // Determine whether its a URL we want native to handle, otherwise continue as normal.
    const shouldMock = urls.find(u => url.startsWith(u));
    if (!shouldMock) {
      return globalFetch(url, options);
    }

    return createPromise('nativeFetch', { url, options })
      .submit()
      .then(({ response, status, statusText }) =>
        Promise.resolve(new Response(response, { status, statusText })),
      );
  };
};
