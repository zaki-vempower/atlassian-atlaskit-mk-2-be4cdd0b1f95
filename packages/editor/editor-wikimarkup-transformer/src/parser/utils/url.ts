/**
 * returns a correctly percent-encoded & sanitized url.
 * Will fallback to using a DOM based implementation if the `URL` class
 * doesn't exist.
 *
 * @param raw a 'raw' url (possibly mixed percent-encoded).
 */
export function decode(raw: string): string {
  if (typeof URL === 'function') {
    const decoded = new URL(raw);
    // IE11 doesn't support the `href` property in the `URL` class.
    // Fallback to the DOM below if it doesn't exist.
    if (decoded.href) {
      return decoded.href;
    }
  }
  // no `URL` class - pollyfill using the dom to parse and
  // decode. This should only be needed for IE11 or lower...
  const anchor = document.createElement('a');
  anchor.href = raw; // The DOM will parses even mixed %-encoded segments of a url when set.
  return anchor.href;
}
