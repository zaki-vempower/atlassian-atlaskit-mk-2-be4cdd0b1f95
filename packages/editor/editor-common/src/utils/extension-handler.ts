import { Extension, ExtensionHandler } from '../extensions/types';

export function getExtensionRenderer<T>(
  extensionHandler: Extension<T> | ExtensionHandler<T>,
): ExtensionHandler<T> {
  if (typeof extensionHandler === 'object') {
    return extensionHandler.render;
  } else {
    return extensionHandler;
  }
}
