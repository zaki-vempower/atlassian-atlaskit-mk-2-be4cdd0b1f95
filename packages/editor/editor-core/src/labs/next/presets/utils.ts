import { EditorPresetProps, PluginsPreset } from './types';
import { EditorPlugin } from '../../../types';
import { compose } from '../../../utils';
import { ProviderFactory } from '@atlaskit/editor-common/provider-factory';

export const removeExcludes = (excludes: EditorPresetProps['excludes']) => (
  plugins: EditorPlugin[],
) => {
  if (excludes) {
    return plugins.filter(plugin => !excludes.has(plugin.name));
  }
  return plugins;
};

export type ExperimentalPluginMap = Map<string, EditorPlugin>;
export const enableExperimental = (
  experimental: EditorPresetProps['experimental'],
  experimentalPluginMap: ExperimentalPluginMap,
) => (plugins: EditorPlugin[]) => {
  if (experimental && experimental.length) {
    experimental.map(pluginName => {
      const plugin = experimentalPluginMap.get(pluginName);
      if (plugin) {
        plugins.push(plugin);
      }
    });
  }
  return plugins;
};

export const processPluginPreset = (preset: PluginsPreset) => {
  const cache = new Map();
  preset.forEach(pluginEntry => {
    if (Array.isArray(pluginEntry)) {
      const [fn, options] = pluginEntry;
      cache.set(fn, options);
    } else {
      /**
       * Prevent usage of same plugin twice without override.
       * [
       *  plugin1,
       *  [plugin1, { option1: value }],
       *  plugin1, // This will throw
       * ]
       */
      if (cache.has(pluginEntry) && cache.get(pluginEntry) === undefined) {
        throw new Error(`${pluginEntry} is already included!`);
      }
      cache.set(pluginEntry, undefined);
    }
  });

  let plugins: Array<EditorPlugin> = [];
  cache.forEach((options, fn) => {
    plugins.push(fn(options));
  });

  return plugins;
};

export const addExcludesFromProviderFactory = (
  providerFactory: ProviderFactory,
  excludes: EditorPresetProps['excludes'] = new Set(),
) => {
  // TODO: Should I make this function pure?
  if (!providerFactory.hasProvider('mentionProvider')) {
    excludes.add('mention');
  }
  if (!providerFactory.hasProvider('emojiProvider')) {
    excludes.add('emoji');
  }
  if (!providerFactory.hasProvider('macroProvider')) {
    excludes.add('macro');
  }
  if (!providerFactory.hasProvider('autoformattingProvider')) {
    excludes.add('customAutoformat');
  }

  return excludes;
};

export const getPluginsFromPreset = (
  preset: PluginsPreset,
  excludes?: EditorPresetProps['excludes'],
  experimental?: EditorPresetProps['experimental'],
): Array<EditorPlugin> => {
  const experimentalMap: ExperimentalPluginMap = new Map();

  return compose(
    enableExperimental(experimental, experimentalMap),
    removeExcludes(excludes),
  )(processPluginPreset(preset));
};
