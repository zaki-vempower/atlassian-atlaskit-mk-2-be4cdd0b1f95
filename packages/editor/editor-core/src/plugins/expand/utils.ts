import { EditorState, Selection } from 'prosemirror-state';
import {
  findSelectedNodeOfType,
  findParentNodeOfType,
} from 'prosemirror-utils';
import { Slice, Schema, Node as PMNode, Fragment } from 'prosemirror-model';
import { mapChildren } from '../../utils/slice';

export const findExpand = (
  state: EditorState,
  selection?: Selection<any> | null,
) => {
  const { expand, nestedExpand } = state.schema.nodes;
  return (
    findSelectedNodeOfType([expand, nestedExpand])(
      selection || state.selection,
    ) ||
    findParentNodeOfType([expand, nestedExpand])(selection || state.selection)
  );
};

export const transformSliceNestedExpandToExpand = (
  slice: Slice,
  schema: Schema,
): Slice => {
  const { expand, nestedExpand } = schema.nodes;
  const children = [] as PMNode[];

  mapChildren(slice.content, (node: PMNode) => {
    if (node.type === nestedExpand) {
      children.push(expand.createChecked(node.attrs, node.content, node.marks));
    } else {
      children.push(node);
    }
  });

  return new Slice(
    Fragment.fromArray(children),
    slice.openStart,
    slice.openEnd,
  );
};
