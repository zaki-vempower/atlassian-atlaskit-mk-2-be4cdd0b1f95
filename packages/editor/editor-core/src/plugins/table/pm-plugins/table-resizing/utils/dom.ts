import { EditorState } from 'prosemirror-state';
import { EditorView } from 'prosemirror-view';
import { TableCssClassName as ClassName } from '../../../types';
import { getPluginState as getMainPluginState } from '../../main';
import { containsClassName, closestElement } from '../../../../../utils';
import { updateOverflowShadows } from '../../../nodeviews/TableComponent';

function getHeights(
  children: NodeListOf<HTMLElement>,
): Array<number | undefined> {
  const heights: Array<number | undefined> = [];
  for (let i = 0, count = children.length; i < count; i++) {
    const child: HTMLElement = children[i] as HTMLElement;
    if (child) {
      const rect = child.getBoundingClientRect();
      const height = rect ? rect.height : child.offsetHeight;
      heights[i] = height;
    } else {
      heights[i] = undefined;
    }
  }
  return heights;
}

export const updateControls = (state: EditorState) => {
  const { tableRef } = getMainPluginState(state);
  if (!tableRef) {
    return;
  }
  const tr = tableRef.querySelector('tr');
  if (!tr) {
    return;
  }
  const wrapper = tableRef.parentElement;
  if (!(wrapper && wrapper.parentElement)) {
    return;
  }

  const rows = tableRef.querySelectorAll('tr');
  const rowControls = wrapper.parentElement.querySelectorAll<HTMLElement>(
    `.${ClassName.ROW_CONTROLS_BUTTON_WRAP}`,
  );
  const numberedRows = wrapper.parentElement.querySelectorAll<HTMLElement>(
    ClassName.NUMBERED_COLUMN_BUTTON,
  );

  const rowHeights = getHeights(rows);

  // update rows controls height on resize
  for (let i = 0, count = rowControls.length; i < count; i++) {
    const height = rowHeights[i];
    if (height) {
      rowControls[i].style.height = `${height + 1}px`;

      if (numberedRows.length) {
        numberedRows[i].style.height = `${height + 1}px`;
      }
    }
  }

  updateOverflowShadows(
    wrapper,
    tableRef,
    wrapper.parentElement.querySelector<HTMLElement>(
      `.${ClassName.TABLE_RIGHT_SHADOW}`,
    ),
    wrapper.parentElement.querySelector<HTMLElement>(
      `.${ClassName.TABLE_LEFT_SHADOW}`,
    ),
  );
};

export const isClickNear = (
  event: MouseEvent,
  click: { x: number; y: number },
): boolean => {
  const dx = click.x - event.clientX,
    dy = click.y - event.clientY;
  return dx * dx + dy * dy < 100;
};

export const getResizeCellPos = (
  view: EditorView,
  event: MouseEvent,
  lastColumnResizable: boolean,
): number | null => {
  const target = event.target as HTMLElement;

  if (!containsClassName(target, ClassName.RESIZE_HANDLE_DECORATION)) {
    return null;
  }

  const tableCell = closestElement(target, 'td, th');

  if (!tableCell) {
    return null;
  }

  const cellStartPosition = view.posAtDOM(tableCell, 0);
  return cellStartPosition - 1;
};
