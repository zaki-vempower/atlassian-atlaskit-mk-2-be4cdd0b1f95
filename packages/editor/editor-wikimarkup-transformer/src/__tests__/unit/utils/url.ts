import { decode } from '../../../parser/utils/url';

const raw =
  'https://example.com/test%2C,%20, space?qu ery=some%20text&query=a%20a,b b,c%2Cc#d,e%2Ce f%20g';
const origURL = window.URL;
const origCreateElement = document.createElement;

describe(`Decoding url`, () => {
  afterEach(() => {
    // @ts-ignore
    window.URL = origURL;
    document.createElement = origCreateElement;
  });

  it(`URL class does not exist`, () => {
    const mockCreateElement = jest.fn();
    // @ts-ignore
    window.URL = undefined;
    // @ts-ignore
    document.createElement = (...args: any) => {
      mockCreateElement();
      return origCreateElement.apply(document, args);
    };
    expect(decode(raw)).toMatchSnapshot();
    expect(mockCreateElement).toHaveBeenCalledTimes(1);
  });

  it(`URL class does not have a href property`, () => {
    const mockCreateElement = jest.fn();
    const mockCtor = jest.fn();
    // @ts-ignore
    window.URL = function(...args: any) {
      mockCtor();
    };
    // @ts-ignore
    document.createElement = (...args: any) => {
      mockCreateElement();
      return origCreateElement.apply(document, args);
    };
    expect(decode(raw)).toMatchSnapshot();
    expect(mockCreateElement).toHaveBeenCalledTimes(1);
    expect(mockCtor).toHaveBeenCalledTimes(1);
  });

  it(`URL class is used if it exists`, () => {
    const mockCreateElement = jest.fn();
    const mockCtor = jest.fn();
    // @ts-ignore
    window.URL = function(...args: any) {
      mockCtor();
      return new origURL(args);
    };
    // @ts-ignore
    document.createElement = (...args: any) => {
      mockCreateElement();
      return origCreateElement.apply(document, args);
    };
    expect(decode(raw)).toMatchSnapshot();
    expect(mockCreateElement).toHaveBeenCalledTimes(0);
    expect(mockCtor).toHaveBeenCalledTimes(1);
  });
});
