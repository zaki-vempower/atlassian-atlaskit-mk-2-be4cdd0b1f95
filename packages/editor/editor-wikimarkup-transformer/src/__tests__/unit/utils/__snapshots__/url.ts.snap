// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Decoding url URL class does not exist 1`] = `"https://example.com/test%2C,%20,%20space?qu%20ery=some%20text&query=a%20a,b%20b,c%2Cc#d,e%2Ce%20f%20g"`;

exports[`Decoding url URL class does not have a href property 1`] = `"https://example.com/test%2C,%20,%20space?qu%20ery=some%20text&query=a%20a,b%20b,c%2Cc#d,e%2Ce%20f%20g"`;

exports[`Decoding url URL class is used if it exists 1`] = `"https://example.com/test%2C,%20,%20space?qu%20ery=some%20text&query=a%20a,b%20b,c%2Cc#d,e%2Ce%20f%20g"`;
